R/compute_ss.R

```diff
@@ -25,18 +27,23 @@
 #'
 compute_suff_stat = function(X, y, standardize = FALSE) {
   y_mean = mean(y)
-  y = y - y_mean
-  is.sparse = !is.matrix(X)
-  X = set_X_attributes(as.matrix(X),center=TRUE,scale = standardize)
-  X = t((t(X) - attr(X,"scaled:center"))/attr(X,"scaled:scale"))
+  y   = y - y_mean
+  n   = nrow(X)
+  mu  = colMeans(X)
+  s   = compute_colSds(X)
   XtX = crossprod(X)
-  if(is.sparse)
-    XtX = as(XtX,"dgCMatrix")
-  Xty = c(y %*% X)
-  n = length(y)
+  XtX = as.matrix(XtX)    
+  XtX = XtX - n*tcrossprod(mu)
```

Comment:
Thanks Gao for going through at the changes carefully.
Yes it is technically true that `crossprod(X)` will result in a sparse matrix when `X` is sparse.
But the next step will always create a dense matrix since `tcrossprod(mu)` is a dense matrix. 
So there is really no benefit here to having `XtX` be sparse.
(Additionally the cross product is often very dense even when X is very sparse.)